File: src/kyu_5/the_hunger_games_zoo_disaster/python/solution.py
EATERS = {"antelope": {"grass"},
          "big-fish": {"little-fish"},
          "bug":      {"leaves"},
          "bear":     {"big-fish", "bug", "chicken", "cow", "leaves", "sheep"},
          "chicken":  {"bug"},
          "cow":      {"grass"},
          "fox":      {"chicken", "sheep"},
          "giraffe":  {"leaves"},
          "lion":     {"antelope", "cow"},
          "panda":    {"leaves"},
          "sheep":    {"grass"}}


def who_eats_who(zoo):

    ansLst, zooLst, n = [zoo], zoo.split(","), 0

    while n < len(zooLst):
        while n > 0 and zooLst[n-1] in EATERS.get(zooLst[n], set()):                            # Eats on its left
            ansLst.append("{} eats {}".format(zooLst[n], zooLst.pop(n-1)))
            n -= 2

        while n >= 0 and n != len(zooLst)-1 and zooLst[n+1] in EATERS.get(zooLst[n], set()):    # Eats on its right
            ansLst.append("{} eats {}".format(zooLst[n], zooLst.pop(n+1)))

        n += 1     # Nothing to eat, step forward

    return ansLst + [','.join(zooLst)]

File: src/kyu_5/the_hunger_games_zoo_disaster/csharp/solution.cs
using System;
using System.Collections.Generic;

public class Dinglemouse {

  private static Dictionary<string, string[]> foodChain = new Dictionary<string, string[]>() {
    {"antelope", new string[] {"grass"}},
    {"big-fish", new string[] {"little-fish"}},
    {"bug", new string[] {"leaves"}},
    {"bear", new string[] {"big-fish", "bug", "chicken", "cow", "leaves", "sheep"}},
    {"chicken", new string[] {"bug"}},
    {"cow", new string[] {"grass"}},
    {"fox", new string[] {"chicken", "sheep"}},
    {"giraffe", new string[] {"leaves"}},
    {"lion", new string[] {"antelope", "cow"}},
    {"panda", new string[] {"leaves"}},
    {"sheep", new string[] {"grass"}},
  };

  public static string[] WhoEatsWho(string zoo) {
    if(zoo == "")
      return new string[1]{""};
    List<string> wew = new List<string>(new string[] {zoo});
    List<string> animals = new List<string>(zoo.Split(','));
    int i = 0;
    int eat = -1;
    while(i < animals.Count) {
      eat = Dinglemouse.Eat(i, animals);
      if(eat > -1) {
        wew.Add(animals[i] + " eats " +  animals[eat]);
        animals.RemoveAt(eat);
        i = 0;
      } else {
        i += 1;
      }
    }
    wew.Add(string.Join(",", animals.ToArray()));
    return wew.ToArray();
  }

  private static int Eat(int index, List<string> animals) {
    string animal = animals[index];
    try {
      string[] eats = Dinglemouse.foodChain[animal];
      if(index > 0 && Array.IndexOf(eats, animals[index-1]) > -1)
        return index - 1;
      if(index < animals.Count-1 && Array.IndexOf(eats, animals[index+1]) > -1)
        return index + 1;
    } catch(KeyNotFoundException) {
    }
    return -1;
  }
}